rsocket-core/src/main/java/io/rsocket/metadata/security/AuthMetadataFlyweight.java

```diff
@@ -0,0 +1,338 @@
+package io.rsocket.metadata.security;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.buffer.ByteBufUtil;
+import io.netty.buffer.Unpooled;
+import io.netty.util.CharsetUtil;
+import io.rsocket.buffer.TupleByteBuf;
+import io.rsocket.util.CharByteBufUtil;
+
+public class AuthMetadataFlyweight {
+
+  static final int STREAM_METADATA_KNOWN_MASK = 0x80; // 1000 0000
+  static final byte STREAM_METADATA_LENGTH_MASK = 0x7F; // 0111 1111
+
+  static final int USERNAME_BYTES_LENGTH = 1;
+  static final int AUTH_TYPE_ID_LENGTH = 1;
+
+  static final char[] EMPTY_CHARS_ARRAY = new char[0];
+
+  private AuthMetadataFlyweight() {}
+
+  /**
+   * Encode a Authentication CompositeMetadata payload using custom authentication type
+   *
+   * @throws IllegalArgumentException if customAuthType is non US_ASCII string if customAuthType is
+   *     empty string if customAuthType is has length greater than 128 bytes
+   * @param allocator the {@link ByteBufAllocator} to use to create intermediate buffers as needed.
+   * @param customAuthType the custom mime type to encode.
+   * @param metadata the metadata value to encode.
+   */
+  public static ByteBuf encodeMetadata(
+      ByteBufAllocator allocator, String customAuthType, ByteBuf metadata) {
+
+    int actualASCIILength = ByteBufUtil.utf8Bytes(customAuthType);
+    if (actualASCIILength != customAuthType.length()) {
+      metadata.release();
+      throw new IllegalArgumentException("custom auth type must be US_ASCII characters only");
+    }
+    if (actualASCIILength < 1 || actualASCIILength > 128) {
+      metadata.release();
+      throw new IllegalArgumentException(
+          "custom auth type must have a strictly positive length that fits on 7 unsigned bits, ie 1-128");
+    }
+
+    int capacity = 1 + actualASCIILength;
+    ByteBuf headerBuffer = allocator.buffer(capacity, capacity);
+    // encoded length is one less than actual length, since 0 is never a valid length, which gives
```

Comment:
Is this worthwhile doing?

Reply:
Kind of pattern used in composite metadata impl. Since spec mentions that max length is 128 then it is the only way to make it. If we are not going to encode it in such a way it will not be possible to distinguish size from encoded wellknownauth type. (0-127 for auth types masked with 0x80 `1000 0000` which means that value overflows) so if one is going to use 128 length for encoded custom auth type then it will be identified as -1&0x7F == 0 well-known mime type